game/screen_songs.cc

```diff
@@ -336,12 +340,46 @@ void ScreenSongs::draw() {
 		theme->song.draw(window, oss_song.str());
 		theme->order.draw(window, oss_order.str());
 		drawInstruments(Dimensions(1.0f).fixedHeight(0.09f).right(0.45f).screenTop(0.02f));
-		theme->hiscores.draw(window, oss_hiscore.str());
+		theme->hiscores.draw(window, hiscore);
 	}
 	// Menus on top of everything
 	if (m_menu.isOpen()) drawMenu();
 }
 
+std::string ScreenSongs::getHighScoreText() const {
+	auto const scores = m_database.queryPerSongHiscore(m_songs.currentPtr());
+	auto const datetimeFormat = config["game/datetime_format"].so();
+	auto const maxLines = 8;
+
+	// Reorder hiscores by track / score
+	std::map<std::string, std::multiset<HiscoreItem>> scoresByTrack;
+	for (auto const& hi: scores)
+		scoresByTrack[hi.track].insert(hi);
+
+	auto stream = std::stringstream();
+	auto n = 0;
+	for (auto const& score: scoresByTrack) {
+		stream << score.first << ":\n";
+		for (auto const& hi: score.second) {
+			stream.width(10);
+			stream << std::right << hi.score<< " \t";
+			stream.width(25);
+			stream << std::left << m_database.getPlayers().lookup(hi.playerid).value_or("Unknown player Id " + std::to_string(hi.playerid));
+
+			if(hi.unixtime.count()) {
+				stream << " \t" << format(hi.unixtime, datetimeFormat);
+			}
+
+			stream << "\n";
+		}
+		stream << "\n";
+		if(++n == maxLines)
+			break;
+	}
+
+	return stream.str();
```

Comment:
I guess we can use range based loops and lambda's here something along the lines of:
```
auto stream = std::stringstream();
int n = 0;
const int maxLines = 10;

for (const auto& [track, scores] : scoresByTrack) {
    stream << track << ":\n";
    for (const auto& score : scores) {
        auto formatScore = [](const auto& score) {
            std::ostringstream ss;
            ss << std::right << std::setw(10) << score << " \t";
            return ss.str();
        };
        auto formatPlayer = [this](const auto& playerId) {
            const auto playerName = m_database.getPlayers().lookup(playerId);
            return std::left << std::setw(25) << (playerName ? *playerName : "Unknown player Id " + std::to_string(playerId));
        };
        stream << formatScore(score.score) << formatPlayer(score.playerid);
        if (score.unixtime.count()) {
            stream << " \t" << format(score.unixtime, datetimeFormat);
        }
        stream << "\n";
        if (++n == maxLines) {
            break;
        }
    }
    stream << "\n";
    if (n == maxLines) {
        break;
    }
}
```

Reply:
Yes I use lambdas now, but a little different as suggested.

Reply:
Instead of using stream.width(n) you can also inline it within the stream itself with `std::setw(n)` i think to remove the need of `stream.width(10);` and `stream.width(25);`
```
auto const scoreFormatter = [](auto& stream, auto const& score){
		stream.width(10);
		stream << std::right << score << " \t";
	};
	auto const playerFormatter = [this](auto& stream, auto const& playerId) {
		const auto player = m_database.getPlayers().lookup(playerId);

		stream.width(25);
		stream << std::left << player.value_or("Unknown player Id " + std::to_string(playerId));
	};
	auto const timeFormatter = [datetimeFormat](auto& stream, auto const& unixtime){
		if(unixtime.count()) {
			stream << " \t" << format(unixtime, datetimeFormat);
		}
	};
```

Reply:
Using stream manipulators now.